## Customer Churn Prediction using Logistic Regression

### Notebook 01 — Data Exploration


### Introduction

Machine Learning projects begin with understanding the dataset before building any predictive model.

In this notebook, we will perform **Data Exploration**, which helps us understand the overall structure of the dataset, identify available features, inspect data types, detect missing values, and examine the target variable.

This exploration provides a solid foundation for data cleaning, feature engineering, and model training in the upcoming notebooks.


### Learning Objectives

By the end of this notebook, we will be able to:

- Load the dataset successfully
- Understand the business problem
- Explore the dataset structure
- Inspect available features
- Examine data types
- Identify missing values
- Detect duplicate records
- Understand the target variable
- Prepare the dataset for preprocessing

### Step 1 — Import Required Libraries

Before working with the dataset, we first import the necessary Python libraries.

These libraries help us perform data manipulation, numerical computations, and suppress unnecessary warnings to keep the notebook clean and readable.

In [1]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np

### Step 2 — Load the Dataset

The next step is to load the customer churn dataset into a Pandas DataFrame.

A DataFrame is a two-dimensional data structure that allows us to efficiently store, manipulate, and analyze tabular data.

In [2]:
df = pd.read_excel(r"D:\machine learning\machine_learning\logistic_regression_project\churn_prediction\data\raw _data\Telco_customer_churn.xlsx")

### Step 3 — Preview the Dataset

After loading the dataset, it is always recommended to inspect the first few records.

This helps us verify that the dataset has been loaded correctly and provides an initial understanding of the available features and their values.

In [4]:
df.head(10)

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,...,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,...,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,...,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,...,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,...,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices
5,4190-MFLUW,1,United States,California,Los Angeles,90020,"34.066367, -118.309868",34.066367,-118.309868,Female,...,Month-to-month,No,Credit card (automatic),55.20,528.35,Yes,1,78,5925,Competitor offered higher download speeds
6,8779-QRDMV,1,United States,California,Los Angeles,90022,"34.02381, -118.156582",34.023810,-118.156582,Male,...,Month-to-month,Yes,Electronic check,39.65,39.65,Yes,1,100,5433,Competitor offered more data
7,1066-JKSGK,1,United States,California,Los Angeles,90024,"34.066303, -118.435479",34.066303,-118.435479,Male,...,Month-to-month,No,Mailed check,20.15,20.15,Yes,1,92,4832,Competitor made better offer
8,6467-CHFZW,1,United States,California,Los Angeles,90028,"34.099869, -118.326843",34.099869,-118.326843,Male,...,Month-to-month,Yes,Electronic check,99.35,4749.15,Yes,1,77,5789,Competitor had better devices
9,8665-UTDHZ,1,United States,California,Los Angeles,90029,"34.089953, -118.294824",34.089953,-118.294824,Male,...,Month-to-month,No,Electronic check,30.20,30.2,Yes,1,97,2915,Competitor had better devices


### Step 4 — Display the Last Five Records

Viewing the last few rows helps ensure that the dataset has been loaded completely and allows us to inspect how the data ends.

Sometimes, missing values or unexpected entries appear toward the end of a dataset, making this step useful for validation.

In [5]:
df.tail(5)

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,...,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
7038,2569-WGERO,1,United States,California,Landers,92285,"34.341737, -116.539416",34.341737,-116.539416,Female,...,Two year,Yes,Bank transfer (automatic),21.15,1419.4,No,0,45,5306,NaN
7039,6840-RESVB,1,United States,California,Adelanto,92301,"34.667815, -117.536183",34.667815,-117.536183,Male,...,One year,Yes,Mailed check,84.80,1990.5,No,0,59,2140,NaN
7040,2234-XADUH,1,United States,California,Amboy,92304,"34.559882, -115.637164",34.559882,-115.637164,Female,...,One year,Yes,Credit card (automatic),103.20,7362.9,No,0,71,5560,NaN
7041,4801-JZAZL,1,United States,California,Angelus Oaks,92305,"34.1678, -116.86433",34.167800,-116.864330,Female,...,Month-to-month,Yes,Electronic check,29.60,346.45,No,0,59,2793,NaN
7042,3186-AJIEK,1,United States,California,Apple Valley,92308,"34.424926, -117.184503",34.424926,-117.184503,Male,...,Two year,Yes,Bank transfer (automatic),105.65,6844.5,No,0,38,5097,NaN


### Step 5 — Check Dataset Dimensions

Understanding the size of the dataset is an important part of data exploration.

The `shape` attribute returns:

- Number of rows (customer records)
- Number of columns (features)

This information helps estimate the amount of available data for training and analysis.

In [7]:
df.shape

(7043, 33)

In [8]:
features_of_dataset = df.shape[1] - 1 
print(f"Number of features in the dataset: {features_of_dataset}")

Number of features in the dataset: 32


### Observation

The dataset dimensions tell us:

- in the data set i can see like there was 7043 rows and 33 column.
- there was 32 features in the this dataset. 
- i can cleary see there was more then 7k rows and for the linear and logistic need to 1k to proper training so it means this was enough - for now 

### Step 6 — Explore Feature Names

Before preprocessing the dataset, it is important to understand all available columns.

Displaying the feature names helps us identify:

- Customer identifiers
- Demographic information
- Service-related features
- Billing information
- Contract details
- Target variable
- Potentially unnecessary columns

In [10]:
df.columns

Index(['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code',
       'Lat Long', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen',
       'Partner', 'Dependents', 'Tenure Months', 'Phone Service',
       'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method',
       'Monthly Charges', 'Total Charges', 'Churn Label', 'Churn Value',
       'Churn Score', 'CLTV', 'Churn Reason'],
      dtype='object')

### Observation

there was some columns that can be impact on the model after training like noise and can be problem of overfiting 
the columns such as customner_id  count  lat  long churn label 